src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,311 @@
+
+/-- Copy of a `cusp_form` with a new `to_fun` equal to the old one. Useful to fix
+definitional equalities. -/
+protected def cusp_form.copy (f : cusp_form Γ k) (f' : ℍ → ℂ) (h : f' = ⇑f) :
+  cusp_form Γ k :=
+{ to_fun := f',
+  slash_action_eq' := h.symm ▸ f.slash_action_eq',
+  hol' := h.symm ▸ f.hol',
+  zero_at_infty' := λ A, h.symm ▸ f.zero_at_infty' A }
+
+end modular_form
+
+namespace modular_form
+
+open slash_invariant_form
+
+variables {F : Type*} {Γ : subgroup SL(2, ℤ)} {k : ℤ}
+
+instance has_add : has_add (modular_form Γ k) :=
+{ add := λ f g,
+  { hol' := f.hol'.add g.hol',
+    bdd_at_infty' := λ A, by simpa using (f.bdd_at_infty' A).add (g.bdd_at_infty' A),
+    .. (f : slash_invariant_form Γ k) + g } }
+
+@[simp] lemma coe_add (f g : modular_form Γ k) : ⇑(f + g) = f + g := rfl
+
+@[simp] lemma add_apply (f g : modular_form Γ k) (z : ℍ) : (f + g) z = f z + g z := rfl
+
+instance has_zero : has_zero (modular_form Γ k) :=
+⟨{ to_fun := 0,
+  slash_action_eq' := slash_action.mul_zero _,
+  hol' := (λ _, mdifferentiable_at_const 𝓘(ℂ, ℂ) 𝓘(ℂ, ℂ)),
+  bdd_at_infty' := λ A, by simpa [slash_action.mul_zero _ ]
+    using (bounded_at_im_infty_subalgebra ℂ).zero_mem' }⟩
+
+section
+variables {α : Type*} [has_smul α ℂ] [is_scalar_tower α ℂ ℂ]
+
+instance has_smul : has_smul α (modular_form Γ k) :=
+⟨ λ c f,
+  { to_fun := c • f,
+    hol' := by simpa using f.hol'.const_smul (c • (1 : ℂ)),
+    bdd_at_infty' := λ A, by simpa using (f.bdd_at_infty' A).const_smul_left (c • (1 : ℂ)),
+     .. c • (f : slash_invariant_form Γ k)}⟩
+
+@[simp] lemma coe_smul (f : (modular_form Γ k)) (n : α) : ⇑(n • f) = n • f := rfl
+@[simp] lemma smul_apply (f : (modular_form Γ k)) (n : α) (z : ℍ) :
+   (n • f) z = n • (f z) := rfl
+end
+
+instance has_neg : has_neg (modular_form Γ k) :=
+⟨λ f,
+  { to_fun := -f,
+    hol' := f.hol'.neg,
+    bdd_at_infty':= λ A, by simpa [-asymptotics.is_O_neg_left] using ((f.bdd_at_infty' A).neg_left),
+    .. -(f : slash_invariant_form Γ k) }⟩
+
+@[simp] lemma coe_neg (f : modular_form Γ k) : ⇑(-f) = -f := rfl
+
+@[simp] lemma neg_apply (f : modular_form Γ k) (z : ℍ) : (-f) z = - (f z) := rfl
+
+instance has_sub : has_sub (modular_form Γ k) :=
+⟨ λ f g, f + -g ⟩
+
+@[simp] lemma coe_sub (f g : (modular_form Γ k)) : ⇑(f - g) = f - g := rfl
+
+@[simp] lemma sub_apply (f g : modular_form Γ k) (z : ℍ) : (f - g) z = f z - g z := rfl
+
+instance : add_comm_group (modular_form Γ k) :=
+fun_like.coe_injective.add_comm_group _ rfl coe_add coe_neg coe_sub coe_smul coe_smul
+
+lemma coe_zero : ⇑(0 : modular_form Γ k) = (0 : ℍ → ℂ) := rfl
+
+/--Additive coercion from `modular_form` to `ℍ → ℂ`. -/
+def coe_hom : (modular_form Γ k) →+ (ℍ → ℂ) :=
+{ to_fun := λ f, f,
+  map_zero' := coe_zero,
+  map_add' := λ _ _, rfl }
+
+lemma coe_hom_injective : function.injective (@coe_hom Γ k) :=
+fun_like.coe_injective
+
+instance : module ℂ (modular_form Γ k) :=
+coe_hom_injective.module ℂ (coe_hom) (λ _ _, rfl)
```

Comment:
```suggestion
function.injective.module ℂ (coe_hom) fun_like.coe_injective (λ _ _, rfl)
```
Should work here, then you can get rid of `coe_hom_injective` (which we tend not to have elsewhere)